src/cpp/core/include/core/r_util/RActiveSession.hpp

```diff
@@ -0,0 +1,492 @@
+   {
+      return readProperty(kRVersionHome);
+   }
+
+   void setRVersion(const std::string& rVersion,
+                    const std::string& rVersionHome,
+                    const std::string& rVersionLabel = "")
+   {
+         writeProperty(kRVersion, rVersion);
+         writeProperty(kRVersionHome, rVersionHome);
+         writeProperty(kRVersionLabel, rVersionLabel);
+   }
+
+   // historical note: this will be displayed as the session name
+   std::string label()
+   {
+      return readProperty(kLabel);
+   }
+
+   // historical note: this will be displayed as the session name
+   void setLabel(const std::string& label)
+   {
+      writeProperty(kLabel, label);
+   }
+
+   void beginSession(const std::string& rVersion,
+                     const std::string& rVersionHome,
+                     const std::string& rVersionLabel = "")
+   {
+      setLastUsed();
+      setRunning(true);
+      setRVersion(rVersion, rVersionHome, rVersionLabel);
+   }
+
+   void endSession()
+   {
+      setLastUsed();
+      setRunning(false);
+      setExecuting(false);
+   }
+
+   uintmax_t suspendSize()
+   {
+      FilePath suspendPath = scratchPath_.completePath("suspended-session-data");
+      if (!suspendPath.exists())
+         return 0;
+
+      return suspendPath.getSizeRecursive();
+   }
+
+   core::Error destroy()
+   {
+      if (!empty())
+         return storage_->destroy();
+      else
+         return Success();
+   }
+
+   bool validate(const FilePath& userHomePath,
+                 bool projectSharingEnabled) const
+   {
+      // ensure the scratch path and properties paths exist
+      if (!scratchPath_.exists() || !propertiesPath_.exists())
+         return false;
+
+      // ensure the properties are there
+      if (project().empty() || workingDir().empty() || (lastUsed() == 0))
+          return false;
+
+      // for projects validate that the base directory still exists
+      std::string theProject = project();
+      if (theProject != kProjectNone)
+      {
+         FilePath projectDir = FilePath::resolveAliasedPath(theProject,
+                                                            userHomePath);
+         if (!projectDir.exists())
+            return false;
+
+        // check for project file
+        FilePath projectPath = r_util::projectFromDirectory(projectDir);
+        if (!projectPath.exists())
+           return false;
+
+        // if we got this far the scope is valid, do one final check for
+        // trying to open a shared project if sharing is disabled
+        if (!projectSharingEnabled &&
+            r_util::isSharedPath(projectPath.getAbsolutePath(), userHomePath))
+           return false;
+      }
+
+      // validated!
+      return true;
+   }
+   
+   bool operator>(const ActiveSession& rhs) const
+   {
+      if (sortConditions_.executing_ == rhs.sortConditions_.executing_)
+      {
+         if (sortConditions_.running_ == rhs.sortConditions_.running_)
+         {
+
```

Comment:
Why did you remove this test?  I don't know that it's important but maybe lastUsed is 0 and this will mean the list order is not consistent which can create weird bugs.

```            
        if (sortConditions_.lastUsed_ == rhs.sortConditions_.lastUsed_)
               return id() > rhs.id();
````

Reply:
Good catch.